File: 218_The_Skyline_Problem/218.cpp
class Solution {
public:
    vector<pair<int, int>> getSkyline(vector<vector<int>>& buildings) {
        vector<pair<int, int>> ans;
        //get all possible points
        unordered_set<int> x_set;
        for(auto &building : buildings){
            x_set.insert(building[0]);
            x_set.insert(building[1]);
        }
        vector<int> x_list(x_set.begin(), x_set.end());
        std::sort(x_list.begin(), x_list.end());
        //iterate through all possible xs'
        priority_queue<pair<int,int>> h_r;
        int itr = 0;
        for(auto &x : x_list){
            //push all building started to heap
            while(itr < buildings.size()){
                if(buildings[itr][0] <= x){
                    h_r.push(make_pair(buildings[itr][2], buildings[itr][1]));
                    ++itr;
                }
                else
                    break;
            }
            //pop from top to ensure the highest building in heap still exist
            while(!h_r.empty()){
                if(h_r.top().second <= x)
                    h_r.pop();
                else
                    break;
            }
            //insert only when height changes
            int h = h_r.empty() ? 0 : h_r.top().first;
            if(ans.size() == 0 || h != ans.back().second)
                ans.push_back(make_pair(x, h));
        }
        return ans;
    }
};

File: 218_The_Skyline_Problem/218.py
class Solution:
    def getSkyline(self, buildings):
        """
        :type buildings: List[List[int]]
        :rtype: List[List[int]]
        """
        ans = []
        from queue import PriorityQueue
        Q = PriorityQueue()
        points = sorted(list(set([b[0] for b in buildings]+[b[1] for b in buildings])))
        # itr points to the leftmost building not been considered
        itr = 0
        # p is the place where we are looking at
        for p in points:
            # add all buildings into PriorityQueue
            while itr < len(buildings) and buildings[itr][0] <= p:
                height = buildings[itr][2]
                rx = buildings[itr][1]
                Q.put((-height, rx))
                itr += 1
            # ensure the highest building still exist
            while not Q.empty():
                h, r = Q.get()
                h = -h
                if r <= p:
                    continue
                else:
                    Q.put((-h, r))
                    break
            # if Q is empty, h should be 0
            h = 0 if Q.empty() else h
            # insert the point into answer if the previous is not the same height
            if len(ans) == 0 or ans[-1][1] != h:
                ans.append([p, h])
        return ans